File: interview/isSubStructure.go
package interview

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSubStructure(A *TreeNode, B *TreeNode) bool {
	if A == nil || B == nil {
		return false
	}
	var order func(na, nb *TreeNode) bool
	order = func(na, nb *TreeNode) bool {
		if nb == nil {
			return true
		}
		if na == nil {
			return false
		}
		if na.Val == nb.Val {
			if order(na.Left, nb.Left) && order(na.Right, nb.Right) {
				return true
			}
		}
		if order(na.Left, nb) || order(na.Right, nb) {
			return true
		}
		return false
	}
	return order(A, B)
}

File: 剑指offer/IsSubStructure.java
import struct.TreeNode.TreeNode;

/**
 * @author Zexho
 * @date 2022/3/4 9:10 AM
 */
public class IsSubStructure {
    public boolean isSubStructure(TreeNode A, TreeNode B) {
        if (A == null) return false;
        return isSub(A, B) || isSubStructure(A.left, B) || isSubStructure(A.right, B);
    }

    public boolean isSub(TreeNode A, TreeNode B) {
        if (A == B || B == null) return true;
        else if (A == null || A.val != B.val) return false;
        return isSub(A.left, B.left) && isSub(A.right, B.right);
    }

}
